File: maven-plugin/src/main/java/TargetDependencyService.java
import model.TargetConfiguration;
import model.TargetDependency;
import org.apache.maven.plugin.logging.Log;
import org.apache.maven.project.MavenProject;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Service responsible for calculating target dependencies in Maven projects.
 * Handles phase dependencies, cross-module dependencies, and goal dependencies.
 */
public class TargetDependencyService {

    private final Log log;
    private final boolean verbose;
    private final ExecutionPlanAnalysisService executionPlanAnalysisService;

    public TargetDependencyService(Log log, boolean verbose, ExecutionPlanAnalysisService executionPlanAnalysisService) {
        this.log = log;
        this.verbose = verbose;
        this.executionPlanAnalysisService = executionPlanAnalysisService;
    }

    /**
     * Calculate dependencies for a goal target.
     * Goals now depend on other goals, not on phases.
     */
    public List<Object> calculateGoalDependencies(MavenProject project, String executionPhase,
                                                  String targetName, List<MavenProject> actualDependencies) {
        List<Object> dependencies = new ArrayList<>();

        String effectivePhase = executionPhase;
        if (effectivePhase == null || effectivePhase.isEmpty() || effectivePhase.startsWith("${")) {
            effectivePhase = executionPlanAnalysisService.findPhaseForGoal(project, ExecutionPlanAnalysisService.extractGoalFromTargetName(targetName));
        }

        if (effectivePhase != null && !effectivePhase.isEmpty()) {
            // Add goal-to-goal dependencies based on Maven lifecycle ordering (same project)
            List<String> precedingGoals = getPrecedingGoalsInLifecycle(project, effectivePhase);
            for (String goal : precedingGoals) {
                dependencies.add(goal); // Simple string dependency for same project
            }

            // Add cross-module dependencies using object form for better precision
            if (!actualDependencies.isEmpty()) {
                List<String> dependentProjects = new ArrayList<>();
                for (MavenProject depProject : actualDependencies) {
                    dependentProjects.add(MavenUtils.formatProjectKey(depProject));
                }
                
                // Create object dependency for cross-module goals in the same phase
                TargetDependency crossModuleDep = new TargetDependency(effectivePhase, dependentProjects);
                dependencies.add(crossModuleDep);
            }
        }

        return dependencies;
    }

    /**
     * Calculate dependencies for a phase target
     */
    public List<Object> calculatePhaseDependencies(String phase, Map<String, TargetConfiguration> allTargets,
                                                   MavenProject project, List<MavenProject> reactorProjects) {
        List<Object> dependsOn = new ArrayList<>();

        // Add dependencies on all goals that belong to this phase
        List<String> goalsForPhase = getGoalsForPhase(phase, allTargets);
        for (String goal : goalsForPhase) {
            dependsOn.add(goal); // Simple string dependency for goals in same project
        }

        return dependsOn;
    }

    /**
     * Get phase dependencies (preceding phases)
     */
    public List<String> getPhaseDependencies(String phase, MavenProject project) {
        List<String> deps = new ArrayList<>();
        String precedingPhase = getPrecedingPhase(phase, project);
        if (precedingPhase != null) {
            deps.add(precedingPhase);
        }
        return deps;
    }

    /**
     * Get all goals that belong to a specific phase
     */
    public List<String> getGoalsForPhase(String phase, Map<String, TargetConfiguration> allTargets) {
        List<String> goalsForPhase = new ArrayList<>();

        for (Map.Entry<String, TargetConfiguration> entry : allTargets.entrySet()) {
            String targetName = entry.getKey();
            TargetConfiguration target = entry.getValue();

            if (target.getMetadata() != null &&
                "goal".equals(target.getMetadata().getType()) &&
                phase.equals(target.getMetadata().getPhase())) {
                goalsForPhase.add(targetName);
            }
        }

        return goalsForPhase;
    }




    /**
     * Get all goals from preceding phases in the Maven lifecycle.
     * This ensures goals depend on other goals, not phases.
     */
    public List<String> getPrecedingGoalsInLifecycle(MavenProject project, String currentPhase) {
        List<String> precedingGoals = new ArrayList<>();

        if (currentPhase == null || currentPhase.isEmpty()) {
            return precedingGoals;
        }

        // Get the lifecycle containing this phase
        org.apache.maven.lifecycle.Lifecycle lifecycle = executionPlanAnalysisService.getLifecycleForPhase(currentPhase);
        if (lifecycle == null || lifecycle.getPhases() == null) {
            return precedingGoals;
        }

        List<String> lifecyclePhases = lifecycle.getPhases();
        int currentPhaseIndex = lifecyclePhases.indexOf(currentPhase);

        if (currentPhaseIndex <= 0) {
            return precedingGoals; // No preceding phases
        }

        // Get all goals from all preceding phases using project's plugin configuration
        for (int i = 0; i < currentPhaseIndex; i++) {
            String precedingPhase = lifecyclePhases.get(i);
            List<String> phaseGoals = getCommonGoalsForPhase(precedingPhase, project);
            precedingGoals.addAll(phaseGoals);
        }

        if (verbose && !precedingGoals.isEmpty()) {
            log.debug("Found " + precedingGoals.size() + " preceding goals for phase '" + currentPhase + "': " + precedingGoals);
        }

        return precedingGoals;
    }
    
    /**
     * Get goals for a Maven lifecycle phase by looking up from the project's plugin configuration.
     * This finds all goals from the project that are bound to the specified phase.
     */
    private List<String> getCommonGoalsForPhase(String phase, MavenProject project) {
        List<String> goals = new ArrayList<>();
        
        if (project == null || project.getBuildPlugins() == null) {
            return goals;
        }
        
        // Go through all plugins and their executions to find goals bound to this phase
        for (org.apache.maven.model.Plugin plugin : project.getBuildPlugins()) {
            if (plugin.getExecutions() != null) {
                for (org.apache.maven.model.PluginExecution execution : plugin.getExecutions()) {
                    String executionPhase = execution.getPhase();
                    
                    // Check if this execution is bound to the phase we're looking for
                    if (phase.equals(executionPhase) && execution.getGoals() != null) {
                        for (String goal : execution.getGoals()) {
                            String pluginGoal = ExecutionPlanAnalysisService.getTargetName(plugin.getArtifactId(), goal);
                            goals.add(pluginGoal);
                        }
                    }
                }
            }
        }
        
        return goals;
    }

    /**
     * Get goals from the same phase across actual project dependencies.
     * Returns dependencies in format "nxProjectName:goalName" where nxProjectName is "groupId:artifactId".
     */
    public List<String> getCrossModuleGoalsForPhase(MavenProject currentProject, String phase, List<MavenProject> actualDependencies) {
        List<String> crossModuleGoals = new ArrayList<>();

        if (phase == null || phase.isEmpty() || actualDependencies == null) {
            return crossModuleGoals;
        }

        // For each actual dependency project
        for (MavenProject otherProject : actualDependencies) {
            if (otherProject != null && !otherProject.equals(currentProject)) {
                // Get all goals for this phase in the other project
                List<String> phaseGoals = executionPlanAnalysisService.getGoalsForPhase(otherProject, phase);

                for (String goal : phaseGoals) {
                    if (goal.contains(":")) {
                        // Add with Nx project name to create project:goal dependency
                        String nxProjectName = MavenUtils.formatProjectKey(otherProject);
                        crossModuleGoals.add(nxProjectName + ":" + goal);
                    }
                }
            }
        }

        if (verbose && !crossModuleGoals.isEmpty()) {
            log.debug("Found " + crossModuleGoals.size() + " cross-module project:goal dependencies from actual dependencies for phase '" + phase + "': " + crossModuleGoals);
        }

        return crossModuleGoals;
    }

    /**
     * Get the preceding phase in the Maven lifecycle (supports all lifecycles: default, clean, site)
     */
    public String getPrecedingPhase(String phase, MavenProject project) {
        if (phase == null || phase.isEmpty()) {
            return null;
        }

        // Find which lifecycle contains this phase and get its phases
        org.apache.maven.lifecycle.Lifecycle lifecycle = executionPlanAnalysisService.getLifecycleForPhase(phase);
        String precedingPhase = null;

        if (lifecycle != null && lifecycle.getPhases() != null) {
            precedingPhase = findPrecedingPhaseInLifecycle(phase, lifecycle.getPhases());
        }

        if (precedingPhase != null && verbose) {
            log.info("Found preceding phase: " + precedingPhase);
        }

        return precedingPhase;
    }

    /**
     * Helper method to find preceding phase within a specific lifecycle
     */
    private String findPrecedingPhaseInLifecycle(String phase, List<String> lifecyclePhases) {
        int currentPhaseIndex = lifecyclePhases.indexOf(phase);
        if (currentPhaseIndex > 0) {
            return lifecyclePhases.get(currentPhaseIndex - 1);
        }
        return null;
    }






}

File: maven-plugin/src/main/kotlin/TargetDependencyService.kt
import model.TargetConfiguration
import model.TargetDependency
import org.apache.maven.plugin.logging.Log
import org.apache.maven.project.MavenProject
import org.apache.maven.model.Plugin
import org.apache.maven.model.PluginExecution

/**
 * Service responsible for calculating target dependencies in Maven projects.
 * Handles phase dependencies, cross-module dependencies, and goal dependencies.
 */
class TargetDependencyService(
    private val log: Log?,
    private val verbose: Boolean,
    private val executionPlanAnalysisService: ExecutionPlanAnalysisService
) {

    /**
     * Calculate dependencies for a goal target.
     * Goals now depend on other goals, not on phases.
     */
    fun calculateGoalDependencies(
        project: MavenProject,
        executionPhase: String?,
        targetName: String,
        actualDependencies: List<MavenProject>
    ): List<Any> {
        val dependencies = mutableSetOf<Any>() // Use Set to prevent duplicates

        val effectivePhase = when {
            executionPhase.isNullOrEmpty() || executionPhase.startsWith("\${") -> {
                executionPlanAnalysisService.findPhaseForGoal(
                    project,
                    ExecutionPlanAnalysisService.extractGoalFromTargetName(targetName)
                )
            }
            else -> executionPhase
        }

        if (effectivePhase?.isNotEmpty() == true) {
            // Add goal-to-goal dependencies based on Maven lifecycle ordering (same project)
            val precedingGoals = getPrecedingGoalsInLifecycle(project, effectivePhase)
            dependencies.addAll(precedingGoals) // Simple string dependencies for same project

            // Add cross-module goal dependencies - goals depend on specific goals, not phases
            if (actualDependencies.isNotEmpty()) {
                val crossModuleGoals = getCrossModuleGoalsForPhase(project, effectivePhase, actualDependencies)
                dependencies.addAll(crossModuleGoals) // Direct goal-to-goal dependencies
            }
        }

        return dependencies.toList() // Convert back to list for return
    }

    /**
     * Calculate dependencies for a phase target
     */
    fun calculatePhaseDependencies(
        phase: String,
        allTargets: Map<String, TargetConfiguration>,
        project: MavenProject,
        reactorProjects: List<MavenProject>
    ): List<Any> {
        val dependsOn = mutableSetOf<Any>() // Use Set to prevent duplicates

        // Add dependencies on all goals that belong to this phase
        val goalsForPhase = getGoalsForPhase(phase, allTargets)
        dependsOn.addAll(goalsForPhase) // Simple string dependencies for goals in same project

        return dependsOn.toList() // Convert back to list for return
    }

    /**
     * Get phase dependencies (preceding phases)
     */
    fun getPhaseDependencies(phase: String, project: MavenProject): List<String> {
        val deps = mutableListOf<String>()
        val precedingPhase = getPrecedingPhase(phase, project)
        precedingPhase?.let { deps.add(it) }
        return deps
    }

    /**
     * Get all goals that belong to a specific phase
     */
    fun getGoalsForPhase(phase: String, allTargets: Map<String, TargetConfiguration>): List<String> {
        return allTargets.entries
            .filter { (_, target) ->
                target.metadata?.type == "goal" && target.metadata?.phase == phase
            }
            .map { (targetName, _) -> targetName }
    }

    /**
     * Get all goals from preceding phases in the Maven lifecycle.
     * This ensures goals depend on other goals, not phases.
     * Optimized to use pre-computed execution plan analysis.
     */
    fun getPrecedingGoalsInLifecycle(project: MavenProject, currentPhase: String?): List<String> {
        val precedingGoals = mutableListOf<String>()

        if (currentPhase.isNullOrEmpty()) {
            return precedingGoals
        }

        // Use pre-computed lifecycle information for better performance
        val lifecycle = executionPlanAnalysisService.getLifecycleForPhase(currentPhase)
        if (lifecycle?.phases == null) {
            return precedingGoals
        }

        val lifecyclePhases = lifecycle.phases
        val currentPhaseIndex = lifecyclePhases.indexOf(currentPhase)

        if (currentPhaseIndex <= 0) {
            return precedingGoals // No preceding phases
        }

        // Get all goals from all preceding phases using project's plugin configuration
        for (i in 0 until currentPhaseIndex) {
            val precedingPhase = lifecyclePhases[i]
            val phaseGoals = getCommonGoalsForPhase(precedingPhase, project)
            precedingGoals.addAll(phaseGoals)
        }

        if (verbose && precedingGoals.isNotEmpty()) {
            log?.debug("Found ${precedingGoals.size} preceding goals for phase '$currentPhase': $precedingGoals")
        }

        return precedingGoals
    }
    
    /**
     * Get goals for a Maven lifecycle phase by looking up from the project's plugin configuration.
     * This finds all goals from the project that are bound to the specified phase.
     */
    private fun getCommonGoalsForPhase(phase: String, project: MavenProject?): List<String> {
        val goals = mutableListOf<String>()
        
        if (project?.buildPlugins == null) {
            return goals
        }
        
        // Go through all plugins and their executions to find goals bound to this phase
        project.buildPlugins.forEach { plugin ->
            plugin.executions?.forEach { execution ->
                val executionPhase = execution.phase
                
                // Check if this execution is bound to the phase we're looking for
                if (phase == executionPhase && execution.goals != null) {
                    execution.goals.forEach { goal ->
                        val pluginGoal = ExecutionPlanAnalysisService.getTargetName(plugin.artifactId, goal, execution.id)
                        goals.add(pluginGoal)
                    }
                }
            }
        }
        
        return goals
    }

    /**
     * Get goals from the same phase across actual project dependencies.
     * Returns dependencies in format "nxProjectName:goalName" where nxProjectName is "groupId:artifactId".
     */
    fun getCrossModuleGoalsForPhase(
        currentProject: MavenProject,
        phase: String?,
        actualDependencies: List<MavenProject>?
    ): List<String> {
        val crossModuleGoals = mutableListOf<String>()

        if (phase.isNullOrEmpty() || actualDependencies == null) {
            return crossModuleGoals
        }

        // For each actual dependency project
        actualDependencies
            .filter { it != currentProject }
            .forEach { otherProject ->
                // Get all goals for this phase in the other project
                val phaseGoals = executionPlanAnalysisService.getGoalsForPhase(otherProject, phase)

                phaseGoals
                    .filter { it.contains(":") }
                    .forEach { goal ->
                        // Add with Nx project name to create project:goal dependency
                        val nxProjectName = MavenUtils.formatProjectKey(otherProject)
                        crossModuleGoals.add("$nxProjectName:$goal")
                    }
            }

        if (verbose && crossModuleGoals.isNotEmpty()) {
            log?.debug("Found ${crossModuleGoals.size} cross-module project:goal dependencies from actual dependencies for phase '$phase': $crossModuleGoals")
        }

        return crossModuleGoals
    }

    /**
     * Get the preceding phase in the Maven lifecycle (supports all lifecycles: default, clean, site)
     */
    fun getPrecedingPhase(phase: String?, project: MavenProject): String? {
        if (phase.isNullOrEmpty()) {
            return null
        }

        // Find which lifecycle contains this phase and get its phases
        val lifecycle = executionPlanAnalysisService.getLifecycleForPhase(phase)
        val precedingPhase = lifecycle?.phases?.let { phases ->
            findPrecedingPhaseInLifecycle(phase, phases)
        }

        if (precedingPhase != null && verbose) {
            log?.info("Found preceding phase: $precedingPhase")
        }

        return precedingPhase
    }

    /**
     * Helper method to find preceding phase within a specific lifecycle
     */
    private fun findPrecedingPhaseInLifecycle(phase: String, lifecyclePhases: List<String>): String? {
        val currentPhaseIndex = lifecyclePhases.indexOf(phase)
        return if (currentPhaseIndex > 0) {
            lifecyclePhases[currentPhaseIndex - 1]
        } else {
            null
        }
    }
}